### ISSUE TITLE
Dimension with zero size raises Reshape error  

### ISSUE THREAD

**Issue by david-macleod** [microsoft/onnxruntime#8421]
**Describe the bug**
When indexing into a tensor to update (`ScatterND`), an implicit reshape is called on the `updates` tensor to add an extra dimension, this Reshape causes an error to be raised when the final dim in the `updates` tensor has size zero.

```
onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running Reshape node. Name:'Reshape_70' Status Message: 
/onnxruntime/onnxruntime/core/providers/cpu/tensor/reshape_helper.h:27 onnxruntime::ReshapeHelper::ReshapeHelper(const onnxruntime::TensorShape&, std::vector<long int>&, bool) i < input_shape.NumDimensions() was false. 
The dimension with value zero exceeds the dimension size of the input tensor.
```

It is very clear why this error is being raised [here](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/reshape_helper.h#L25), in my example below the input_shape is `(2, 0)`, the requested shape is `(1, 2, 0)` and it doesn't like the fact that a dim of zero size exists in a position > number of dims in input_shape. My question is **why does this check exist**? This works fine in PyTorch (presumably resulting in a no-op).

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: 1.8.0
- Python version: 3.8

**To Reproduce**
```py
    import torch
    import onnxruntime as ort

    class Model(torch.nn.Module):

        def forward(self, x, y):
            x[0,:,:y.shape[1]] = y
            return x

    model = torch.jit.script(Model())
    x, y = torch.randn(2,2,2), torch.randn(2,1)
    output = model(x,y)
    torch.onnx.export(model, (x,y), "tmp.onnx", input_names=["x","y"], dynamic_axes={"y":[1]},
    example_outputs=output, opset_version=13, verbose=True)

    sess = ort.InferenceSession("tmp.onnx")
    # THIS WORKS
    y= torch.randn(2,1)
    print(sess.run(None, {"x": x.numpy(), "y": y.numpy()}))
    # THS FAILS
    y= torch.randn(2,0)
    print(sess.run(None, {"x": x.numpy(), "y": y.numpy()}))
```

**Expected behavior**
No error to be raised, the reshape works correctly as per PyTorch.

My current solution is to add conditionals that check the `numel()` of the incoming tensor > 0 before peforming the update, but ideally I shouldn't have to do this. Can I rewrite the update in a cleaner way to avoid the error?

**Comment by hariharans29** [comment]
`Why does this check exist ?`

Reshape's [spec](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Reshape-13) up until opset-13 dictated that a value of `0` in the shape input meant that the corresponding dimension value in the input shape was preserved in the output shape. Obviously, a value of `0` in an index greater than the rank of the input is an error by this spec.

The opset-14 [spec](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Reshape) allows doing what you are trying to do (i.e.) treat a `0` in the shape input as a `0` itself. But this requires setting the attribute `allowzero` to `1` (by default it is 0).

TL;DR:
To achieve what you are trying to do, you need an opset-14 model exported from Torch and the Reshape node must have the attribute `allowzero` set to `1`. If that is not the case, please check with the Torch exporter group as to how to enable this. If that is the already case and you still hit this runtime check, it is a bug in ORT. Please attach a simplistic model to help repro this.

**Comment by hariharans29** [comment]
Closing as this the expected behavior for opset-13. If you encounter something similar with opset-14 with the right attribute set, please re-open.
